File: mobile/test/basket_run_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:trolley_scout/api_models.dart';
import 'package:trolley_scout/basket_run.dart';

void main() {
  test('groups quantities and source-backed totals by retailer', () {
    final basket = Basket(
      items: [
        _item('bread', 'shoprite', 'Shoprite', 2, 4000, 600),
        _item('milk', 'checkers', 'Checkers', 1, 2300, 200),
        _item('eggs', 'shoprite', 'Shoprite', 1, null, null),
      ],
      summary: const BasketSummary(
        itemCount: 4,
        knownPriceItemCount: 3,
        savingsCents: 800,
        totalCents: 6300,
      ),
    );

    final stops = buildBasketStoreStops(basket);
    expect(stops, hasLength(2));
    expect(stops.first.retailerId, 'checkers');
    expect(stops.first.totalCents, 2300);
    expect(stops.last.retailerId, 'shoprite');
    expect(stops.last.itemCount, 3);
    expect(stops.last.knownPriceItemCount, 2);
    expect(stops.last.totalCents, 4000);
    expect(stops.last.savingsCents, 600);
  });

  test('formats a shareable list with store stops, links, and honest totals',
      () {
    final basket = Basket(
      items: [
        _item('bread', 'shoprite', 'Shoprite', 2, 4000, 600),
        _item('eggs', 'shoprite', 'Shoprite', 1, null, null),
      ],
      summary: const BasketSummary(
        itemCount: 3,
        knownPriceItemCount: 2,
        savingsCents: 600,
        totalCents: 4000,
      ),
    );

    final text = formatBasketShareText(
      basket,
      formatMoney: (cents) => 'R${(cents / 100).toStringAsFixed(2)}',
    );

    expect(text, contains('1. Shoprite (R40.00 known subtotal)'));
    expect(text, contains('• 2 × bread: R40.00'));
    expect(text, contains('https://example.test/bread'));
    expect(text, contains('• 1 × eggs: price not found'));
    expect(text, contains('Savings found: R6.00'));
  });
}

BasketItem _item(
  String id,
  String retailerId,
  String retailerName,
  int quantity,
  int? linePriceCents,
  int? lineSavingCents,
) =>
    BasketItem(
      id: id,
      savedDealId: 'saved-$id',
      quantity: quantity,
      deal: SavedDeal(
        id: 'deal-$id',
        retailerId: retailerId,
        retailerName: retailerName,
        sourceLabel: 'Weekly offers',
        sourceUrl: 'https://example.test/$retailerId',
        productUrl: 'https://example.test/$id',
        title: id,
        capturedAt: '2026-08-01T10:00:00.000Z',
        evidenceText: id,
        savedAt: '2026-08-01T10:00:00.000Z',
      ),
      linePriceCents: linePriceCents,
      lineSavingCents: lineSavingCents,
    );

File: src/services/basketRun.test.ts
import { describe, expect, it } from "vitest";
import type { Basket, BasketItem, SavedDeal } from "../types";
import { buildBasketStoreStops, formatBasketShareText } from "./basketRun";

function deal(id: string, retailerId: string, retailerName: string): SavedDeal {
  return {
    capturedAt: "2026-08-01T10:00:00.000Z",
    evidenceText: id,
    id,
    productUrl: `https://example.test/${id}`,
    retailerId: retailerId as SavedDeal["retailerId"],
    retailerName,
    savedAt: "2026-08-01T10:00:00.000Z",
    sourceLabel: "Weekly offers",
    sourceUrl: `https://example.test/${retailerId}`,
    title: id,
  };
}

function item(
  id: string,
  retailerId: string,
  retailerName: string,
  quantity: number,
  linePriceCents?: number,
  lineSavingCents?: number,
): BasketItem {
  return {
    addedAt: "2026-08-01T10:00:00.000Z",
    deal: deal(`deal-${id}`, retailerId, retailerName),
    id,
    linePriceCents,
    lineSavingCents,
    quantity,
    savedDealId: `saved-${id}`,
    updatedAt: "2026-08-01T10:00:00.000Z",
  };
}

describe("buildBasketStoreStops", () => {
  it("groups quantities and source-backed totals by retailer", () => {
    const basket: Basket = {
      items: [
        item("bread", "shoprite", "Shoprite", 2, 4000, 600),
        item("milk", "checkers", "Checkers", 1, 2300, 200),
        item("eggs", "shoprite", "Shoprite", 1),
      ],
      summary: {
        itemCount: 4,
        knownPriceItemCount: 3,
        savingsCents: 800,
        totalCents: 6300,
      },
    };

    expect(buildBasketStoreStops(basket)).toEqual([
      expect.objectContaining({
        itemCount: 1,
        knownPriceItemCount: 1,
        retailerId: "checkers",
        savingsCents: 200,
        totalCents: 2300,
      }),
      expect.objectContaining({
        itemCount: 3,
        knownPriceItemCount: 2,
        retailerId: "shoprite",
        savingsCents: 600,
        totalCents: 4000,
      }),
    ]);
  });

  it("formats a shareable list with store stops, links, and honest totals", () => {
    const basket: Basket = {
      items: [
        item("bread", "shoprite", "Shoprite", 2, 4000, 600),
        item("eggs", "shoprite", "Shoprite", 1),
      ],
      summary: {
        itemCount: 3,
        knownPriceItemCount: 2,
        savingsCents: 600,
        totalCents: 4000,
      },
    };

    const text = formatBasketShareText(basket, (cents) => `R${(cents / 100).toFixed(2)}`);

    expect(text).toContain("1. Shoprite (R40.00 known subtotal)");
    expect(text).toContain("• 2 × deal-bread: R40.00");
    expect(text).toContain("https://example.test/deal-bread");
    expect(text).toContain("• 1 × deal-eggs: price not found");
    expect(text).toContain("Savings found: R6.00");
  });
});
